model/storage/ConsumerStorage.php

```diff
@@ -81,13 +81,44 @@ public function setConsumerToken(\core_kernel_classes_Resource $consumer, Access
      */
     public function getToken($tokenHash)
     {
-        if ($this->getCache()->exists($tokenHash)) {
-            $token = new AccessToken(json_decode($this->getCache()->get($tokenHash), true));
-        } else {
-            $encodedToken = $this->getConsumerByTokenHash($tokenHash)->getOnePropertyValue($this->getProperty(self::CONSUMER_TOKEN));
-            $token = new AccessToken(json_decode($encodedToken, true));
-            $this->getCache()->set($token->getToken(), $encodedToken);
+        $cache = $this->getCache();
+
+        if ($cache->exists($tokenHash)) {
+            $decodedToken = json_decode($this->getCache()->get($tokenHash), true);
+
+            if (is_array($decodedToken)) {
+                return new AccessToken($decodedToken);
+            }
+
+            $this->logWarning(
+                sprintf(
+                    'The token %s contains an invalid JSON: %s',
+                    substr($tokenHash, 0, 10) . '...',
+                    json_last_error_msg()
+                )
+            );
+        }
+
+        $encodedToken = $this->getConsumerByTokenHash($tokenHash)
+            ->getOnePropertyValue($this->getProperty(self::CONSUMER_TOKEN));
+
+        $decodedToken = json_decode($encodedToken, true);
+
+        if (!is_array($decodedToken)) {
+            $errorMessage = sprintf(
+                'The token %s contains an invalid JSON: %s',
+                substr($tokenHash, 0, 10) . '...',
+                json_last_error_msg()
+            );
+
+            $this->logError($errorMessage);
```

Comment:
in L:93 this same message maybe we can put decoding of token to private method and reuse it?